### DISCUSSION TITLE
Type AppMain does not conform to protocol App

### DISCUSSION THREAD

**Discussion by davidakoontz** [skiptools/skip#188] [Q&A] [answered]
I'm a bit wrapped up and bound by all this... stuff beyond the capabilities of a simple cave man.

The error on my Skip AppMain is

Error 1:  Inheritance from non-protocol type 'PainlessPasswordApp'
Error 2:  Type 'AppMain' does not conform to protocol 'App'

The file:  PainlessPasswordAppMain.swift

`
import SwiftUI
import PainlessPassword

/// The entry point to the app simply loads the App implementation from SPM module.
@main struct AppMain: App, PainlessPasswordApp {
}
`

Any clue to resolve this - I've been deleting Build Folder & DerivedData folder...  

Any other ideas?

**Comment by davidakoontz** [comment]
I started over - back at the beginning... and had to remove some of the changes I'd made...
at some point I thought (WRONGLY) that making the PainlessPasswordApp NOT implement the App protocol AND
Changing from an extension the concrete implementation might solve some of the problems ...  it just messed more things up
So I backed out those.... Incorrect assumptions and changes.

Logging the BONEHEAD idea here... 

`#if !SKIP
public protocol PainlessPasswordApp : App {
}

/// The entry point to the PainlessPassword app.
/// The concrete implementation is in the PainlessPasswordApp module.
public extension PainlessPasswordApp {

    var body: some Scene {
        // 'environmentObject' is only available in iOS 17.0
        WindowGroup {
            RootView()
        }
        // Skip does not like the WindowGroup.environmentObject()
        // so we put it in first View ContentView()
    }   // body Scene
    
    
}
#endif
`

**Comment by marcprux** [comment, marked as answer]
Yes, you'll want to keep that protocol implementation. All the code that is transpiled in Skip is done in a SwiftPM module, but since Xcode requires at least one top-level target (that isn't a SwiftPM module), our `skip init` command creates an `AppMain` shim that looks like:

```
import SwiftUI
import HelloSkip

/// The entry point to the app simply loads the App implementation from SwiftPM module.
@main struct AppMain: App, HelloSkipApp {
}
```

This implementation is left empty, because all the logic is defined in a the `HelloSkipApp` protocol in the `HelloSkip` SwiftPM module, which will look like this:

```
public protocol HelloSkipApp : App {
}

/// The entry point to the HelloSkip app.
/// The concrete implementation is in the HelloSkipApp module.
public extension HelloSkipApp {
    var body: some Scene {
        WindowGroup {
            RootView()
        }
    }
}
```

This allows you to define your app logic in the SwiftPM module in a way that can be shared with the Android side.

The short answer, though, is that it is probably best not to mess around with the auto-generated app scaffolding, and concentrate more on staying within the guidelines for the best experience.

**Comment by davidakoontz** [comment]
Right - good advice - stay within the guides!  

I'm still pondering how the Android App is launched...  but for now just happy I can inject my AppState ObservedObject into the Env.  and it works over on the Kotlin side also!
